{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%load_ext watermark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%watermark -a Schmelling,Nicolas -u -d -v -p matplotlib,numpy,pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "Any comments and suggestions or questions?     \n",
    "Please feel free to contact me via [twitter](https://twitter.com/DerSchmelling) or [email](mailto:Nicolas.Schmelling@hhu.de).\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Distribution of circadian clock proteins ##\n",
    "\n",
    "In the previous [IPython notebook](1_KaiABC_BLAST_Data_Collection_and_Perprocessing.ipynb) we collected the data and preformed some preprocessing steps. In this notebook we continued to preprocess the data such that it fits the desired analyses. Afterwards we visualized the data in three heatmaps containing the percent identity of the reciprocal BLAST and the count of protein homologs in cyanobacteria.\n",
    "\n",
    "### Preprocessing ###\n",
    "\n",
    "1. Grouping the genera in the four groups according to their taxonomy. Groups are Cyanobacteria, Proteobacteria, Archaea, and Other Bacteria. Assign a numeric identifier to each group for coloring.\n",
    "2. Calculating the identity mean for each protein and genus. A counter of each genus to record the number of individual organisms per genus will be created.\n",
    "3. Merging the the dataframes for each protein into a single dataframe containing information about protein identity for each protein and genus.\n",
    "4. Creating another dataframe containing the homologous copy number for each protein in cyanobacteria."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "import matplotlib as mlp\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import ticker\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pd.set_option('mode.chained_assignment',None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "'''\n",
    "The function will group the genera in to four groups according to\n",
    "their taxonomy.Groups are Cyanobacteria, Proteobacteria, \n",
    "Archaea, and Other Bacteria. The function will also assign a numeric \n",
    "identifier to each group for later coloring.\n",
    "'''\n",
    "def categories_taxonomy(df):\n",
    "    \n",
    "    count = 0\n",
    "    \n",
    "    df['colormap'] = np.zeros(len(df.index))\n",
    "    \n",
    "    for i in list(df.taxonomy):\n",
    "        if 'cyanobacteria' in str(i).lower():\n",
    "            df.taxonomy[count] = 1\n",
    "            df['colormap'][count] = 3\n",
    "            count += 1\n",
    "        elif 'proteobacteria' in str(i).lower():\n",
    "            df.taxonomy[count] = 3\n",
    "            df['colormap'][count] = 2\n",
    "            count += 1\n",
    "        elif 'archaea' in str(i).lower():\n",
    "            df.taxonomy[count] = 4\n",
    "            df['colormap'][count] = 1\n",
    "            count += 1\n",
    "        else:\n",
    "            df.taxonomy[count] = 5\n",
    "            df['colormap'][count] = 9\n",
    "            count += 1\n",
    "    \n",
    "    return df\n",
    "\n",
    "'''\n",
    "The function will first call the function categories_taxonomy() and \n",
    "selects only cyanobacteria. Next a counter to record the number protein \n",
    "homologs of individual organisms will be created.\n",
    "'''\n",
    "def create_count_df(path, prot_name, df_merge, how='left'):\n",
    "    \n",
    "    df = pd.read_csv(path)\n",
    "    df = categories_taxonomy(df)\n",
    "    df = df[df.taxonomy == 1]\n",
    "    \n",
    "    leng = prot_name + '_length'\n",
    "    coun = prot_name + '_count'\n",
    "    \n",
    "    df = df[['name',leng]]\n",
    "    df = df[df[leng] > 0].drop_duplicates()\n",
    "    df[coun] = df.groupby(['name']).transform('count')\n",
    "    df = df[['name',coun]].drop_duplicates()\n",
    "    \n",
    "    df_new = pd.merge(df_merge, df, how=how)\n",
    "    \n",
    "    return df_new\n",
    "\n",
    "'''\n",
    "The function will first call the function create_count_df() and further \n",
    "calculate the identity mean for each protein and genus. It will also\n",
    "remove all special characters, change the genus name of all organisms\n",
    "labeled as 'Candidatus', and create a counter for each genus \n",
    "to record the number of individual organisms per genus. \n",
    "'''\n",
    "def prepare_df(df, prot):\n",
    "    \n",
    "    df = categories_taxonomy(df)\n",
    "    df = df[['name','genus','taxonomy','colormap','identity','genome_id']]\n",
    "    \n",
    "    count = 0\n",
    "    for genus in df['genus']:\n",
    "        df['genus'][count] = re.sub('[^a-zA-Z0-9 \\n\\.]', '', genus)\n",
    "        if genus == 'Candidatus':\n",
    "            df['genus'][count] = df['name'][count].split(' ',2)[1]\n",
    "        count += 1\n",
    "        \n",
    "    df = df.groupby(['name','genus','taxonomy','colormap','genome_id']).mean()\n",
    "    df = df.reset_index()\n",
    "    df['%s_count'%prot] = df.drop_duplicates(subset=['name'])\\\n",
    "                                            [['name','genus']]\\\n",
    "                                            .groupby(['genus'])\\\n",
    "                                            .transform('count')\n",
    "    df = df[['name','genus','taxonomy','colormap','identity','%s_count'%prot]]\n",
    "    df.columns = ['name', 'genus', 'taxonomy','colormap',\\\n",
    "                  '%s_identity'%prot,'%s_count'%prot]\n",
    "    df = df.groupby(['name','genus','taxonomy']).mean()\n",
    "    df = df.reset_index()\n",
    "    \n",
    "    return df\n",
    "\n",
    "'''\n",
    "Read the CSV of a given protein and run the prepare_df() function.\n",
    "'''\n",
    "def process_df(prot):\n",
    "    \n",
    "    df = pd.read_csv('../data/%s.csv'%prot)\n",
    "    df = prepare_df(df, prot)\n",
    "    \n",
    "    return df\n",
    "\n",
    "'''\n",
    "The function will read a given CSV file, runs the categories_taxonomy function\n",
    "and selects only cyanobacteria. Next a counter to record the number protein \n",
    "homologs of individual organisms will be created. Last the resulting\n",
    "two column dataframe will be merged with a given existing dataframe.\n",
    "'''\n",
    "def create_count_df(path, prot_name, df_merge, how='left'):\n",
    "    \n",
    "    df = pd.read_csv(path)\n",
    "    df = categories_taxonomy(df)\n",
    "    df = df[df.taxonomy == 1]\n",
    "    \n",
    "    leng = prot_name + '_length'\n",
    "    coun = prot_name + '_count'\n",
    "    \n",
    "    df = df[['name','length']]\n",
    "    df.columns = ['name',leng]\n",
    "    df = df[df[leng] > 0].drop_duplicates()\n",
    "    df[coun] = df.groupby(['name']).transform('count')\n",
    "    df = df[['name',coun]].drop_duplicates()\n",
    "    \n",
    "    df_new = pd.merge(df_merge, df, how=how)\n",
    "    \n",
    "    return df_new"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create the sequence identity dataframes ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Run the process_df() function for all Synechococcus proteins\n",
    "for protein in ['kaiA', 'kaiB', 'kaiC', 'cpmA', 'ircA', 'pex', 'cdpA',\n",
    "                'ldpA', 'labA', 'lalA', 'cikA', 'sasA', 'nhtA', 'prkE',\n",
    "                'crm', 'rpaA', 'rpaB']:\n",
    "    globals()[protein] = process_df(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Create a list of genera that are removed later from the dataset,\n",
    "# which were falsely recorded.\n",
    "kaiC_orgs = list(set(kaiC.genus))\n",
    "\n",
    "not_found = ['Oscillatoriales', 'Opitutaceae', 'halophilic',\n",
    "             'Candidatus', 'Aquifex', 'Filomicrobium', 'Ziziphus']\n",
    "\n",
    "orgs = list(set(kaiC_orgs) - set(not_found))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Merge the dataframes of each protein into one dataframe.\n",
    "# First merge the dataframes containing information about KaiA, KaiB and KaiC.\n",
    "# Remove all genera that are not in the organism list created before.\n",
    "kaiC = kaiC[kaiC.genus.isin(orgs)]\n",
    "df_Sy7942 = pd.merge(kaiC,kaiA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, kaiB, how='left')\n",
    "\n",
    "# Merge the df_Sy7942 dataframe with the rest of the protein dataframes.\n",
    "df_Sy7942 = pd.merge(df_Sy7942, cpmA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, ircA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, pex, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, cdpA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, ldpA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, labA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, lalA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, cikA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, sasA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, prkE, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, nhtA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, crm, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, rpaA, how='left')\n",
    "df_Sy7942 = pd.merge(df_Sy7942, rpaB, how='left')\n",
    "\n",
    "df_Sy7942 = df_Sy7942.groupby(['genus','taxonomy']).mean()\n",
    "df_Sy7942 = df_Sy7942.reset_index()\n",
    "\n",
    "df_Sy7942['rank'] = np.zeros(len(df_Sy7942))\n",
    "\n",
    "# Change the genus of a special cyanobacterium\n",
    "id = df_Sy7942[df_Sy7942.genus == 'cyanobacterium'].taxonomy.index[0]\n",
    "df_Sy7942.genus[id] = 'Endosymbiont'\n",
    "\n",
    "# Reduce the number of columns down to the one needed for the heatmap.\n",
    "df_Sy7942 = df_Sy7942[['genus', 'taxonomy', 'colormap', 'rank',\n",
    "                       'kaiA_identity', 'kaiB_identity', 'kaiC_identity',\n",
    "                       'pex_identity', 'ldpA_identity', 'prkE_identity',\n",
    "                       'nhtA_identity', 'ircA_identity', 'cdpA_identity',\n",
    "                       'cikA_identity', 'sasA_identity', 'labA_identity',\n",
    "                       'lalA_identity', 'crm_identity', 'rpaA_identity',\n",
    "                       'rpaB_identity', 'cpmA_identity', 'kaiC_count']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Read the input data for each single protein and run the two function above.\n",
    "for protein in ['kaiB1', 'kaiB2', 'kaiB3', 'kaiC1', 'kaiC2', 'kaiC3']:\n",
    "    globals()[protein] = process_df(protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df_Sy6803 = pd.merge(kaiB1,kaiB2, how='outer')\n",
    "df_Sy6803 = pd.merge(df_Sy6803, kaiB3, how='outer')\n",
    "df_Sy6803 = pd.merge(df_Sy6803, kaiC1, how='outer')\n",
    "df_Sy6803 = pd.merge(df_Sy6803, kaiC2, how='outer')\n",
    "df_Sy6803 = pd.merge(df_Sy6803, kaiC3, how='outer')\n",
    "df_Sy6803 = df_Sy6803.drop_duplicates()\n",
    "df_Sy6803 = df_Sy6803[df_Sy6803.genus.isin(orgs)]\n",
    "\n",
    "df_Sy6803 = df_Sy6803.groupby(['genus','taxonomy']).mean()\n",
    "df_Sy6803 = df_Sy6803.reset_index()\n",
    "\n",
    "df_Sy6803['rank'] = np.zeros(len(df_Sy6803))\n",
    "\n",
    "# Change the genus of a special cyanobacterium\n",
    "id = df_Sy6803[df_Sy6803.genus == 'cyanobacterium'].taxonomy.index[0]\n",
    "df_Sy6803.genus[id] = 'Endosymbiont'\n",
    "\n",
    "# Reduce the number of columns down to the one needed for the heatmap.\n",
    "df_Sy6803 = df_Sy6803[['genus', 'taxonomy', 'colormap', 'rank',\n",
    "                       'kaiB1_identity', 'kaiB2_identity', 'kaiB3_identity',\n",
    "                       'kaiC1_identity', 'kaiC2_identity', 'kaiC3_identity']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Sort dataframe first by the four assigned groups \n",
    "# and then by the KaiC identity.\n",
    "df_Sy7942_sorted = df_Sy7942.sort_values(by=['taxonomy','kaiC_identity'],\\\n",
    "                                         ascending=[True,False])\n",
    "\n",
    "sorterIndex = dict(zip(df_Sy7942_sorted.genus,\\\n",
    "                       range(len(df_Sy7942_sorted.genus))))\n",
    "\n",
    "# Use this order to sort the other dataframe\n",
    "df_Sy6803_sorted = df_Sy6803.copy()\n",
    "df_Sy6803_sorted['rank'] = df_Sy6803_sorted.genus.map(sorterIndex)\n",
    "\n",
    "df_Sy6803_sorted = df_Sy6803_sorted.sort_values(by=['rank'],\\\n",
    "                                                ascending = [True])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create the protein copy number dataframe for cyanobacteria ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Create the count_df dataframe. First KaiA, KaiB1 and KaiC1. \n",
    "# Then add the other Kai proteins from Synechocystis \n",
    "# and the rest from Synechococcus. This dataframe will be used \n",
    "# later for Fisher's exact test.\n",
    "count_df = pd.read_csv('../data/kaiA.csv')\n",
    "count_df = categories_taxonomy(count_df)\n",
    "count_df = count_df[count_df.taxonomy == 1]\n",
    "count_df = count_df[['name','length']]\n",
    "count_df.columns = ['name','kaiA_length']\n",
    "count_df = count_df[count_df.kaiA_length > 0].drop_duplicates()\n",
    "count_df['kaiA_count'] = count_df.groupby(['name']).transform('count')\n",
    "count_df = count_df[['name','kaiA_count']].drop_duplicates()\n",
    " \n",
    "protein_list = ['kaiB1','kaiC1','kaiB2','kaiC2','kaiB3','kaiC3',\n",
    "                'cpmA','ircA','pex','nhtA','prkE','cdpA','ldpA',\n",
    "                'cikA','labA','sasA','lalA','rpaA','rpaB','crm']\n",
    "\n",
    "for protein in protein_list[:2]:\n",
    "    count_df = create_count_df('../data/%s.csv' %protein,\n",
    "                               protein, count_df, how='outer')\n",
    "    \n",
    "for protein in protein_list[2:6]:\n",
    "    count_df = create_count_df('../data/%s.csv' %protein,\n",
    "                               protein, count_df, how='left')\n",
    "    \n",
    "for protein in protein_list[6:]:\n",
    "    count_df = create_count_df('../data/%s.csv' %protein,\n",
    "                               protein, count_df, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Save the count_df dataframe, because we need it later for\n",
    "# the pairwise Fisher's exact test.\n",
    "count_df.to_csv('../clock_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Readn the previously created clock_counts.csv and sort\n",
    "# the rows descending by portein counts\n",
    "df_count = pd.read_csv('../clock_counts.csv', index_col='Unnamed: 0')\n",
    "\n",
    "df_count = df_count.sort_values(by=['kaiA_count', 'kaiC1_count', 'kaiB1_count',\n",
    "                                    'kaiB3_count', 'kaiC3_count', 'kaiB2_count',\n",
    "                                    'kaiC2_count', 'cpmA_count', 'ircA_count',\n",
    "                                    'pex_count', 'nhtA_count', 'prkE_count',\n",
    "                                    'cdpA_count', 'ldpA_count', 'cikA_count',\n",
    "                                    'labA_count', 'sasA_count', 'lalA_count',\n",
    "                                    'rpaA_count', 'rpaB_count', 'crm_count'],\n",
    "                                ascending=[False, False, False, False, False,\n",
    "                                           False, False, False, False, False,\n",
    "                                           False, False, False, False, False,\n",
    "                                           False, False, False, False, False,\n",
    "                                           False])\n",
    "\n",
    "df_count = df_count[['name', 'kaiA_count', 'kaiB1_count', 'kaiC1_count',\n",
    "                     'kaiB2_count', 'kaiC2_count', 'kaiB3_count',\n",
    "                     'kaiC3_count', 'pex_count', 'ldpA_count', 'prkE_count',\n",
    "                     'nhtA_count', 'ircA_count', 'cdpA_count', 'cikA_count', \n",
    "                     'sasA_count', 'labA_count', 'lalA_count', 'crm_count', \n",
    "                     'rpaA_count', 'rpaB_count', 'cpmA_count']]\n",
    "\n",
    "# Overwrite the name of two special cyanobacteria\n",
    "name = df_count['name']\n",
    "\n",
    "name[name.str.contains('Candidatus')] = 'Atelocyanobacterium ' \\\n",
    "                                        'thalassa isolate ALOHA '\\\n",
    "                                        '(cyanobacteria)'\n",
    "name[name.str.contains('endosymbiont')] = 'Endosymbiont '\\\n",
    "                                          '(cyanobacteria)'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Heatmap with sequence identities and copy numbers ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# The function takes three dataframes and a file name,\n",
    "# creates a heatmap, and saves it.\n",
    "def heatmap(df1, df2, df3, name):\n",
    "    \n",
    "    # Create the empty figure with 4.35 inch x 8.75 inch in size\n",
    "    f = plt.figure(figsize=(4.35,8.75))\n",
    "    \n",
    "    # Create a special grid von subplots\n",
    "    plots = gridspec.GridSpec(6, 6, wspace=0.05, hspace=0.1,\n",
    "                              width_ratios=[0.15, 2.8, 1.15, 1.05, 0.6, 3.2],\n",
    "                              height_ratios=[0.6, 0.01, 0.01, 0.4, 0.01, 0.01])\n",
    "    \n",
    "    hfont = {'fontname':'Arial'}\n",
    "    \n",
    "    mlp.rcParams['lines.linewidth'] = 3\n",
    "    \n",
    "    # Remove the plot frame lines.\n",
    "    def clean_axis(ax):\n",
    "        for sp in ax.spines.values():\n",
    "            sp.set_visible(False)\n",
    "    \n",
    "############ 1st Heatmap Synechococcus proteins ############\n",
    "    \n",
    "    syn_heat = f.add_subplot(plots[:-2,1])\n",
    "    \n",
    "    # Remove the tick marks and place labels at top, bottom, and right  \n",
    "    syn_heat.tick_params(axis=\"both\", which=\"both\", bottom=\"off\", top=\"off\",\n",
    "                    left=\"off\", right=\"off\", labeltop=\"on\", labelbottom=\"on\", \n",
    "                    labelleft=\"off\", labelright=\"on\")\n",
    "    \n",
    "    clean_axis(syn_heat)\n",
    "\n",
    "    # Put the major ticks at the middle of each cell\n",
    "    syn_heat.set_xticks(np.arange(df1.shape[1])+0.5, minor=False)\n",
    "    syn_heat.set_yticks(np.arange(df1.shape[0])+0.5, minor=False)\n",
    "    \n",
    "    # Add x-axis labels\n",
    "    syn_heat.set_xticklabels(['KaiA', 'KaiB', 'KaiC', 'Pex', 'LdpA', 'PrkE',\n",
    "                              'NhtA', 'IrcA', 'CdpA', 'CikA', 'SasA', 'LabA',\n",
    "                              'LalA', 'Crm', 'RpaA', 'RpaB', 'CpmA'],\n",
    "                             fontsize=3.5, rotation=90, **hfont)\n",
    "    \n",
    "    # Create a list with y-labels. Each y-label should contain \n",
    "    # the genus name and the number of underlying organisms.\n",
    "    ylabels = []\n",
    "    \n",
    "    for names, counts in zip(list(df1.genus),list(df1.kaiC_count)):\n",
    "        ylabels.append(str(names)+' ('+str(int(counts))+')')\n",
    "    \n",
    "    syn_heat.set_yticklabels(ylabels, fontsize=2, **hfont)\n",
    "\n",
    "    # Limit the range of the plot to only where the data is.  \n",
    "    # Avoid unnecessary whitespace.\n",
    "    syn_heat.set_ylim(plt.ylim(0, len(df1.index))[::-1]) \n",
    "\n",
    "    # Let's do the actual ploting\n",
    "    heat_syn = plt.pcolor(df1.iloc[:,4:-1], cmap='GnBu',vmin=0, vmax=100)\n",
    "\n",
    "############ 2nd Heatmap with KaiBC123 ############\n",
    "    \n",
    "    bc_heat = f.add_subplot(plots[:-2,3], sharey=syn_heat)\n",
    "    \n",
    "    bc_heat.tick_params(axis=\"both\", which=\"both\", bottom=\"off\", top=\"off\",\n",
    "                    left=\"off\", right=\"off\", labeltop=\"on\", labelbottom=\"on\", \n",
    "                    labelleft=\"off\", labelright=\"off\")\n",
    "\n",
    "    clean_axis(bc_heat)\n",
    "\n",
    "    bc_heat.set_xticks(np.arange(df2.shape[1])+0.5, minor=False)\n",
    "    \n",
    "    bc_heat.set_xticklabels(['KaiB1','KaiB2','KaiB3','KaiC1','KaiC2','KaiC3'],\n",
    "                        fontsize=3.5, rotation=90, **hfont)    \n",
    "\n",
    "    heat_bc = plt.pcolor(df2.iloc[:,4:], cmap='GnBu',vmin=0, vmax=100) \n",
    "    \n",
    "############ Colorbar for the 1st and 2nd Heatmap ############\n",
    "    \n",
    "    bar_heat = f.add_subplot(plots[-1,1:4])\n",
    "    \n",
    "    # Modify colorbar\n",
    "    cbar = f.colorbar(heat_syn, bar_heat, orientation='horizontal')\n",
    "    \n",
    "    tick_locator = ticker.MaxNLocator(nbins=11)\n",
    "    cbar.locator = tick_locator\n",
    "    \n",
    "    cbar.update_ticks() \n",
    "    \n",
    "    # Set labels and tick labels\n",
    "    cbar.ax.tick_params(length=1.5, width=0.3)\n",
    "    cbar.ax.set_xticklabels(np.arange(0,101,10), fontsize=3.5, **hfont)\n",
    "    cbar.set_label('Identity [%]', fontsize=4, **hfont)\n",
    "\n",
    "    cbar.outline.set_linewidth(0)\n",
    "    \n",
    "############ Taxonomic Groups (Cyanobacteria, Proteobacteria, Archaea, Other)\n",
    "    \n",
    "    tax_bar = f.add_subplot(plots[:-2,0], sharey=syn_heat)\n",
    "   \n",
    "    tax_bar.tick_params(axis=\"both\", which=\"both\", bottom=\"off\", top=\"off\",\n",
    "                    left=\"off\", right=\"off\", labeltop=\"off\", labelbottom=\"off\", \n",
    "                    labelleft=\"off\", labelright=\"off\")\n",
    "    \n",
    "    clean_axis(tax_bar)\n",
    "    \n",
    "    # Create a colored bar that corresponds to the four groups.\n",
    "    heatmap_taxo = plt.pcolor(df1.iloc[:,2:3], cmap='Set1',vmin=1, vmax=9)\n",
    "    \n",
    "############ 3rd Heatmap Clock counts ############\n",
    "    \n",
    "    count_heat = f.add_subplot(plots[0,5])\n",
    "\n",
    "    count_heat.tick_params(axis=\"both\", which=\"both\", bottom=\"off\", top=\"off\",  \n",
    "                    labelbottom=\"on\", left=\"off\", right=\"off\", labelleft='off',\n",
    "                    labelright='on', labeltop=\"on\")\n",
    "    \n",
    "    clean_axis(count_heat)\n",
    "    \n",
    "    PCC7942 = ['KaiA', 'KaiB1', 'KaiC1', 'KaiB2', 'KaiC2', 'KaiB3', 'KaiC3',\n",
    "               'Pex', 'LdpA', 'PrkE', 'NhtA', 'IrcA', 'CdpA', 'CikA', 'SasA',\n",
    "               'LabA', 'LalA', 'Crm', 'RpaA', 'RpaB', 'CpmA']\n",
    "\n",
    "    names = []\n",
    "\n",
    "    for n in df3.name:\n",
    "        names.append(n.split(' (')[0])\n",
    "\n",
    "    count_heat.set_xticks(np.arange(len(df3.columns[1:]))+0.5, minor=False)\n",
    "    count_heat.set_xticklabels(PCC7942, fontsize=3.5, rotation=90, **hfont)\n",
    "    \n",
    "    count_heat.set_yticks(np.arange(len(df3.index))+0.5, minor=False)\n",
    "    count_heat.set_yticklabels(names, fontsize=2.5, rotation=0, **hfont)\n",
    "    \n",
    "    count_heat.set_xlim(0, 22)\n",
    "    count_heat.set_ylim(0, len(df3.index))\n",
    "\n",
    "    heatmap = plt.pcolor(df3.iloc[:,1:], cmap='Greens', vmax=5, vmin=0)\n",
    "    \n",
    "    count_heat.invert_yaxis()\n",
    "    \n",
    "############ Colorbar for 3rd Heatmap ############\n",
    "    \n",
    "    bar_count = f.add_subplot(plots[2,5])\n",
    "    \n",
    "    cbar = f.colorbar(heatmap, bar_count, orientation='horizontal')\n",
    "    \n",
    "    tick_locator = ticker.MaxNLocator(nbins=6)\n",
    "    cbar.locator = tick_locator\n",
    "    \n",
    "    cbar.update_ticks()\n",
    "\n",
    "    cbar.ax.tick_params(length=1.5, width=0.3)\n",
    "    cbar.ax.set_xticklabels(['0','1','2','3','4','5+'], fontsize=3.5, **hfont)\n",
    "    cbar.set_label('No. of copies', fontsize=4, **hfont)\n",
    "\n",
    "    cbar.outline.set_linewidth(0)\n",
    "    \n",
    "############ Add some text ############\n",
    "    \n",
    "    syn_heat.text(-2,-5,'A',fontsize=8, **hfont)\n",
    "    bc_heat.text(-2,-5,'B',fontsize=8, **hfont)\n",
    "    bc_heat.text(8.5,-5,'C',fontsize=8, **hfont)\n",
    "    \n",
    "    plt.savefig(name, format='eps', bbox_inches='tight', dpi=1000)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "heatmap(df_Sy7942_sorted, df_Sy6803_sorted, df_count,\n",
    "        '../heatmap.eps')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Previous ###\n",
    "\n",
    "+ [Data Collection and Processing](1_KaiABC_BLAST_Data_Collection_and_Perprocessing.ipynb)\n",
    "\n",
    "### Next ###\n",
    "\n",
    "+ [Length Distribution of KaiA, KaiB, KaiC](3_KaiABC_BLAST_Scatterplot.ipynb)\n",
    "+ [Co-Occurence of Circadian Clock Proteins in Cyanobacteria](4_KaiABC_BLAST_FisherTest.ipynb)\n",
    "+ [Additional Analyses](5_KaiABC_BLAST_Other.ipynb)\n",
    "\n",
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
